### PR TITLE
Fix loading toolbar factory twice

### PR DESCRIPTION
**Pull request by fcollonval**
## References

Seen in https://github.com/jupyterlab-contrib/jupyterlab-filesystem-access/pull/32

## Code changes

- Tolerate to create a toolbar factory multiple times from the same plugin (like multiple filebrowser with the same buttons).
- For late plugin registration, we reload all plugin toolbar definitions as for the shortcuts.
- Compare only the toolbar item name to determine if definitions must be merged

## User-facing changes

N/A

## Backwards-incompatible changes

N/A

### CODE REVIEW

**Comment by jtpio** [comment]
Thanks!

Changes look good. Curious if you were able to try it with https://github.com/jupyterlab-contrib/jupyterlab-filesystem-access/pull/32 and the `--extensions-in-dev-mode` flag?

**Comment by fcollonval** [comment]
> Changes look good. Curious if you were able to try it with [jupyterlab-contrib/jupyterlab-filesystem-access#32](https://github.com/jupyterlab-contrib/jupyterlab-filesystem-access/pull/32) and the `--extensions-in-dev-mode` flag?

Yes it works.

**Review by jtpio — APPROVED** [review]
Thanks!

File: packages/apputils/src/toolbar/factory.ts [hunk 1]
```diff
@@ -170,41 +185,34 @@ async function setToolbarItems(
...
-            (a, b) =>
-              (a.rank ?? DEFAULT_TOOLBAR_ITEM_RANK) -
-              (b.rank ?? DEFAULT_TOOLBAR_ITEM_RANK)
-          );
-          transferSettings(newList);
+  if (listenPlugin) {
+    registry.pluginChanged.connect(async (sender, plugin) => {
+      // As the plugin storing the toolbar definition is transformed using
+      // the above definition, if it changes, this means that a request to
+      // reloaded was triggered. Hence the toolbar definitions from the other
```
Comment (afshin) [inline packages/apputils/src/toolbar/factory.ts:192]:
```suggestion
      // reload was triggered. Hence the toolbar definitions from the other
```

File: packages/apputils/src/toolbar/factory.ts [hunk 2]
```diff
@@ -170,41 +185,34 @@ async function setToolbarItems(
...
-              (a.rank ?? DEFAULT_TOOLBAR_ITEM_RANK) -
-              (b.rank ?? DEFAULT_TOOLBAR_ITEM_RANK)
-          );
-          transferSettings(newList);
+  if (listenPlugin) {
+    registry.pluginChanged.connect(async (sender, plugin) => {
+      // As the plugin storing the toolbar definition is transformed using
+      // the above definition, if it changes, this means that a request to
+      // reload was triggered. Hence the toolbar definitions from the other
+      // plugins has been automatically reset during the transform step.
```
Comment (afshin) [inline packages/apputils/src/toolbar/factory.ts:193]:
```suggestion
      // plugins have been automatically reset during the transform step.
```

File: packages/settingregistry/src/settingregistry.ts [hunk 3]
```diff
@@ -1195,39 +1195,18 @@ export namespace SettingRegistry {
...
-          const refIndex = items.findIndex(ref => ref.name === item.name);
-          if (refIndex < 0) {
-            items.push({ ...item });
-          } else {
-            if (warn) {
-              console.warn(`Toolbar item '${item.name}' is duplicated.`);
-            }
-            items[refIndex] = { ...items[refIndex], ...item };
-          }
+      // Name must be unique so it sufficient to only compare it
```
Comment (afshin) [inline packages/settingregistry/src/settingregistry.ts:1198]:
```suggestion
      // Name must be unique so it's sufficient to only compare it
```

**Comment by jtpio** [comment]
@afshin looks like all the comments have been addressed, and this should be good to go now?

**Comment by fcollonval** [comment]
@meeseeksdev please backport to 3.4.x

**Pull request merged**